File: src/test/helpers/golden.ts
// SPDX-License-Identifier: LicenseRef-FSL-1.1-Apache-2.0
/**
 * Golden / characterization fixtures — the cross-language parity oracle for the
 * Go rewrite (#210, #225 item 2).
 *
 * Each golden captures a real Node-backend response (REST body, WS frame, MCP
 * tool output, audit page) with volatile fields replaced by stable placeholders,
 * committed as JSON under `<testdir>/__goldens__/`. The Go server is expected to
 * replay the same request, apply the SAME normalization (documented below and in
 * docs/api/README.md), and diff against the identical file — so a shape/field/
 * status drift fails loudly in either implementation.
 *
 * Normalization contract (keep in sync with the Go side):
 *   - Keys carrying wall-clock or per-run values → placeholder, regardless of
 *     type: createdAt, updatedAt, lastActivityAt, lastUsedAt, builtAt,
 *     authorizedAt, closedAt, resolvedAt, expiresAt, createdAtEpoch → "<TS>".
 *   - challenge, challengeId, token, stepUpToken → "<REDACTED>".
 *   - nextCursor (when non-null) → "<CURSOR>".
 *   - Value patterns anywhere:
 *       sess_<12 hex>                → "sess_<ID>"
 *       bare UUID v1-5               → "<UUID>"
 *       ISO-8601 timestamp           → "<TS>"
 *       SHA256:<b64> fingerprint     → "<FINGERPRINT>" (test SSH key is per-run)
 *   - Any occurrence of the live server origin (http://127.0.0.1:<port>) in a
 *     string → "<BASE_URL>" (ports are per-run). Discovery docs additionally run
 *     against a pinned externalUrl so their bodies are stable on their own.
 *   - Numeric fields equal to a per-run ephemeral port (ssh/app) → "<PORT>"
 *     (e.g. the seeded endpoint's `port` is the test SSH server's random port).
 *
 * Regenerate all fixtures after an intentional contract change:
 *     UPDATE_GOLDENS=1 pnpm test:integration
 */
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { expect } from "vitest";

const UPDATE = process.env.UPDATE_GOLDENS === "1";

const TS_KEYS = new Set([
  "createdAt",
  "updatedAt",
  "lastActivityAt",
  "lastUsedAt",
  "builtAt",
  "authorizedAt",
  "closedAt",
  "resolvedAt",
  "expiresAt",
]);
const REDACT_KEYS = new Set(["challenge", "challengeId", "token", "stepUpToken"]);

const SESSION_ID_RE = /^sess_[0-9a-f]{12}$/;
const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
const ISO_RE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:\d{2})?$/;
const FINGERPRINT_RE = /^SHA256:[A-Za-z0-9+/=]+$/;

export interface NormalizeOptions {
  /** Live server origin(s) to fold to "<BASE_URL>" in string values. */
  baseUrls?: string[];
  /** Per-run ephemeral ports (ssh/app) to fold to "<PORT>" in numeric fields. */
  ports?: number[];
}

function normalizeString(s: string, baseUrls: string[]): string {
  let out = s;
  for (const b of baseUrls) if (b) out = out.split(b).join("<BASE_URL>");
  if (out !== s) return out; // was a URL-bearing string; don't also pattern-swap it
  if (SESSION_ID_RE.test(out)) return "sess_<ID>";
  if (UUID_RE.test(out)) return "<UUID>";
  if (ISO_RE.test(out)) return "<TS>";
  if (FINGERPRINT_RE.test(out)) return "<FINGERPRINT>";
  return out;
}

/** Deep-normalize a captured value into its stable golden form. */
export function normalizeGolden(value: unknown, opts: NormalizeOptions = {}): unknown {
  const baseUrls = opts.baseUrls ?? [];
  const ports = opts.ports ?? [];
  function walk(v: unknown, key?: string): unknown {
    if (key && TS_KEYS.has(key) && v !== null && v !== undefined) return "<TS>";
    if (key && REDACT_KEYS.has(key) && typeof v === "string") return "<REDACTED>";
    if (key === "nextCursor" && typeof v === "string" && v.length > 0) return "<CURSOR>";
    if (typeof v === "number" && ports.includes(v)) return "<PORT>";
    if (typeof v === "string") return normalizeString(v, baseUrls);
    if (Array.isArray(v)) return v.map((item) => walk(item));
    if (v && typeof v === "object") {
      const out: Record<string, unknown> = {};
      for (const [k, val] of Object.entries(v as Record<string, unknown>)) out[k] = walk(val, k);
      return out;
    }
    return v;
  }
  return walk(value);
}

/**
 * Assert `actual` matches the committed golden `<testDirOf(metaUrl)>/__goldens__/<name>.json`,
 * after normalization. With `UPDATE_GOLDENS=1` (or when the file is absent) it
 * (re)writes the fixture instead of asserting. Returns the normalized value.
 */
export function expectGolden(
  metaUrl: string,
  name: string,
  actual: unknown,
  opts: NormalizeOptions = {},
): unknown {
  const dir = join(dirname(fileURLToPath(metaUrl)), "__goldens__");
  const file = join(dir, `${name}.json`);
  const normalized = normalizeGolden(actual, opts);
  const serialized = `${JSON.stringify(normalized, null, 2)}\n`;

  if (UPDATE || !existsSync(file)) {
    mkdirSync(dir, { recursive: true });
    writeFileSync(file, serialized);
    return normalized;
  }
  const golden = JSON.parse(readFileSync(file, "utf8"));
  expect(normalized, `golden mismatch for "${name}" (run UPDATE_GOLDENS=1 to refresh)`).toEqual(
    golden,
  );
  return normalized;
}

File: internal/golden/golden.go
// SPDX-License-Identifier: LicenseRef-FSL-1.1-Apache-2.0
// Package golden is the Go side of the cross-language parity oracle
// (#210/#225): a port of src/test/helpers/golden.ts's normalization. Both
// implementations must fold volatile values to the same placeholders so they
// can diff against the same committed fixtures
// (src/test/integration/__goldens__/*.json). Keep the rule set in lockstep
// with golden.ts — it is the contract, not an implementation detail.
package golden

import (
	"regexp"
	"strings"
)

var tsKeys = map[string]bool{
	"createdAt": true, "updatedAt": true, "lastActivityAt": true,
	"lastUsedAt": true, "builtAt": true, "authorizedAt": true,
	"closedAt": true, "resolvedAt": true, "expiresAt": true,
}

var redactKeys = map[string]bool{
	"challenge": true, "challengeId": true, "token": true, "stepUpToken": true,
}

var (
	sessionIDRe   = regexp.MustCompile(`^sess_[0-9a-f]{12}$`)
	uuidRe        = regexp.MustCompile(`(?i)^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$`)
	isoRe         = regexp.MustCompile(`^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:\d{2})?$`)
	fingerprintRe = regexp.MustCompile(`^SHA256:[A-Za-z0-9+/=]+$`)
)

// Options mirror NormalizeOptions in golden.ts.
type Options struct {
	// BaseURLs are live server origins folded to "<BASE_URL>" in strings.
	BaseURLs []string
	// Ports are per-run ephemeral ports folded to "<PORT>" in a `port` field.
	Ports []float64
}

func normalizeString(s string, baseURLs []string) any {
	out := s
	for _, b := range baseURLs {
		if b != "" {
			out = strings.ReplaceAll(out, b, "<BASE_URL>")
		}
	}
	if out != s {
		return out // was a URL-bearing string; don't also pattern-swap it
	}
	switch {
	case sessionIDRe.MatchString(out):
		return "sess_<ID>"
	case uuidRe.MatchString(out):
		return "<UUID>"
	case isoRe.MatchString(out):
		return "<TS>"
	case fingerprintRe.MatchString(out):
		return "<FINGERPRINT>"
	}
	return out
}

// Normalize deep-normalizes a decoded JSON value (map[string]any / []any /
// string / float64 / bool / nil) into its stable golden form — the exact
// walk of normalizeGolden in golden.ts.
func Normalize(value any, opts Options) any {
	return walk(value, "", opts)
}

func walk(v any, key string, opts Options) any {
	if key != "" && tsKeys[key] && v != nil {
		return "<TS>"
	}
	if key != "" && redactKeys[key] {
		if _, ok := v.(string); ok {
			return "<REDACTED>"
		}
	}
	if key == "nextCursor" {
		if s, ok := v.(string); ok && len(s) > 0 {
			return "<CURSOR>"
		}
	}
	// Scoped to `port` on purpose (see golden.ts): key-agnostic folding would
	// mask any numeric field that happened to equal an ephemeral port.
	if key == "port" {
		if n, ok := v.(float64); ok {
			for _, p := range opts.Ports {
				if p == n {
					return "<PORT>"
				}
			}
		}
	}
	switch t := v.(type) {
	case string:
		return normalizeString(t, opts.BaseURLs)
	case []any:
		out := make([]any, len(t))
		for i, item := range t {
			out[i] = walk(item, "", opts)
		}
		return out
	case map[string]any:
		out := make(map[string]any, len(t))
		for k, val := range t {
			out[k] = walk(val, k, opts)
		}
		return out
	default:
		return v
	}
}
